### ISSUE TITLE
Runtime error with MSSQL@7.0.0-alpha.4

### ISSUE THREAD

**Issue by oranyelik** [knex/knex#4216]
# Environment

Knex version: `0.21.15`
Database + version:
```
Microsoft SQL Server 2016 (SP1) (KB3182545) - 13.0.4001.0 (X64) 
	Oct 28 2016 18:17:30 
	Copyright (c) Microsoft Corporation
	Developer Edition (64-bit) on Windows 10 Pro N 6.3 <X64> (Build 19041: )
```
Database library: `mssql@7.0.0-alpha.4`
OS: `macOS 11.1 (20C69)`

Tagging @smorey2 .

# Bug

1. Runtime error occurs when using the alpha version of MSSQL database library with latest version of knex.

2. Error message:
```
(node:83915) UnhandledPromiseRejectionWarning: Error: Tarn: unsupported option opt.evictionRunIntervalMillis
    at new Pool (/.../knex-mssql7-test/node_modules/tarn/dist/Pool.js:64:23)
    at /.../knex-mssql7-test/node_modules/mssql/lib/base/connection-pool.js:181:21
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

3. Reduced test code:
```
// package.json
{
  "name": "knex-mssql7-test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "node ./index.js"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "knex": "^0.21.15",
    "mssql": "^7.0.0-alpha.4"
  }
}
```
```
// index.js
const knex = require('knex');

const db = knex({
    client: 'mssql',
    connection: {
        host: `...`,
        user: `...`,
        password: `...`,
        database: `...`,
        options: {
            trustServerCertificate: true, // required for my self-signed certificate
        },
    },
});

db.raw(`SELECT @@version`).then(result => console.log('result', result));
```

**Comment by kibertoad** [comment]
Looks like your  pooling configuration is wrong. Can you check `pool` branch of your config? See https://knexjs.org/#Installation-pooling

**Comment by oranyelik** [comment]
Here's some information that appears to be relevant to the issue:

MSSQL@6.3.1 depends on Tarn@1.1.5: https://github.com/tediousjs/node-mssql/blob/6/package.json
MSSQL@7 depends on Tarn@3.0.1: https://github.com/tediousjs/node-mssql/blob/master/package.json

Tarn does not and has never supported the `evictionRunIntervalMillis` option. See: https://github.com/Vincit/tarn.js/blob/master/src/Pool.ts

However, as of Tarn@2.0.0, unsupported options now throw errors. See: https://github.com/Vincit/tarn.js/commit/dc7cf19a37dfe4eb578ac6a79bedefe59e4a13e9

I was planning to submit a PR to fix the issue but it seems that I don't have permission. Knex appears to connect successfully if references to `evictionRunIntervalMillis` are removed entirely from the MSSQL dialect. If you want to add me as a contributor I'd be happy to push up my local working branch. If not, here is the entirety of the changes:

```
knex % git branch
* fix/mssql7
  master
knex % git diff master
diff --git a/lib/constants.js b/lib/constants.js
index 75127dd0..53cf6b0e 100644
--- a/lib/constants.js
+++ b/lib/constants.js
@@ -23,7 +23,6 @@ const POOL_CONFIG_OPTIONS = Object.freeze([
   'fifo',
   'priorityRange',
   'autostart',
-  'evictionRunIntervalMillis',
   'numTestsPerRun',
   'softIdleTimeoutMillis',
   'Promise',
diff --git a/lib/dialects/mssql/index.js b/lib/dialects/mssql/index.js
index a85c98e6..61c4a2e0 100644
--- a/lib/dialects/mssql/index.js
+++ b/lib/dialects/mssql/index.js
@@ -31,7 +31,6 @@ function Client_MSSQL(config = {}) {
     min: 1,
     max: 1,
     idleTimeoutMillis: Number.MAX_SAFE_INTEGER,
-    evictionRunIntervalMillis: 0,
   };
 
   Client.call(this, config);
diff --git a/types/index.d.ts b/types/index.d.ts
index 215d3dc1..ae114760 100644
--- a/types/index.d.ts
+++ b/types/index.d.ts
@@ -1956,7 +1956,6 @@ export declare namespace Knex {
       fifo?: boolean;
       priorityRange?: number;
       autostart?: boolean;
-      evictionRunIntervalMillis?: number;
       numTestsPerRun?: number;
       softIdleTimeoutMillis?: number;
       Promise?: any;
```

**Comment by oranyelik** [comment]
> Looks like your pooling configuration is wrong. Can you check `pool` branch of your config? See https://knexjs.org/#Installation-pooling

I am using the default pooling configuration. The reduced test code provided in the original issue description is the entirety of a codebase that reproduces the issue.

**Comment by oranyelik** [comment]
Note that a similar PR was created and closed in 2019 for unrelated reasons: https://github.com/knex/knex/pull/3044.

**Comment by kibertoad** [comment]
@oranyelik You don't need contributor permissions to create a PR, you can create one from your personal fork of the repo.
I'm fine with dropping support for old node-mssql versions, but it has to be mentioned in `UPGRADING.md`

**Comment by oranyelik** [comment]
@kibertoad To expand on what you said, `node-mssql` changed from depending on `generic-pool@3.6.1` to `tarn@1.1.4` in this commit: https://github.com/tediousjs/node-mssql/commit/a799be574e0558a3053820d97bea27a2ac8e0874#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519. Right after this commit, `node-mssql` transitions from `v5` to `v6.0.0-alpha.1`.

This is significant because generic-pool is what originally supported `evictionRunIntervalMillis`: https://github.com/coopernurse/node-pool#documentation

If we want to maintain knex support for `node-mssql@5`, we must support `evictionRunIntervalMillis`. In its current form, however, supporting this option will not allow us to support `node-mssql@7`.

**Comment by kibertoad** [comment]
@oranyelik Considering that we list `mssql` peer dependency at `^6.3.1` and next knex release is semver major, as I've said, I'm fine with dropping support for older node-mssql versions, as long as it is properly documented. Better yet, we should throw an error from the dialect ourselves if version is too old.
This would be a good place: https://github.com/knex/knex/blob/0bef361ffdc089714d0b5c81fe769ab95a20bf65/lib/dialects/mssql/index.js#L67

**Comment by elhigu** [comment]
I don't see how `tarn` or `generic-pool` that is used by `node-mssql` internally has anything to do with knex. Knex shouldn't pass its pool options to node-mssql's pool, because knex avoids using any pool implementations of DB driver libraries (though node-mssql doesn't allow to omit it completely).

Also we don't want to support any mssql which is still using generic pool. That version had many critical bugs and were leaking connections etc. (I noticed those in stress tests and then wrote monkeypatches to knex side to support it more or less safely)... afterwards node-mssql maintainers started fixing those issues and also started using tarn internally and nowadays we can use newer node-mssql versions without our own patches...

**Comment by elhigu** [comment]
If the old mssql monkeypatch code is still in mssql dialect that should be removed and relay on newer mssql versions only.

**Comment by kibertoad** [comment]
@oranyelik Hotfix was released in 0.21.16, and next semver major will feature a completely redone integration with mssql dialect, dropping support for older versions.
